File: LeetCode75/N_ary_Tree_Preorder_Traversal/N_ary_Tree_Preorder_Traversal.cc
// 589. N-ary Tree Preorder Traversal C++
#include <vector>

// Definition for a Node.
class Node {
public:
    int val;
    std::vector<Node*> children;

    Node() {}

    Node(int _val) {
        val = _val;
    }

    Node(int _val, std::vector<Node*> _children) {
        val = _val;
        children = _children;
    }
};

class Solution {
public:
    std::vector<int> list;
    void traverse(Node* root) {
        if (root == nullptr) return;
        list.emplace_back(root->val);
        for (auto x: root->children) {
            traverse(x);
        }
    }

    std::vector<int> preorder(Node* root) {
        traverse(root);
        return list;
    }
};

File: LeetCode75_LevelOne/N_ary_Tree_Preorder_Traversal/N_ary_Tree_Preorder_Traversal.py
# 589. N-ary Tree Preorder Traversal Python
"""
# Definition for a Node.
class Node(object):
    def __init__(self, val=None, children=None):
        self.val = val
        self.children = children
"""

class Solution(object):
    def preorder(self, root):
        """
        :type root: Node
        :rtype: List[int]
        """
        lst = []
        self.traverse(root, lst)
        return lst
    
    def traverse(self, root, lst):
        if root == None: return

        lst.append(root.val)
        for child in root.children:
            self.traverse(child, lst)
